docs/modules/ROOT/pages/gsn.adoc

```diff
@@ -0,0 +1,77 @@
+
+Unlike regular contract function calls, each relayed call has an additional number of steps it must go through. Most of writing a recipient involves handling these, which are designed to provide enough flexibility to cater to all possible use cases. Basic recipients can however safely ignore most of them, while still being secure and sound.
+
+The OpenZeppelin Contracts provide a number of [tried-and-tested approaches]() for you to use out of the box, but you should still have a basic idea of what's going on under the hood.
+
+==== acceptRelayedCall
+
+First, RelayHub will ask your recipient contract if it wants to receive a relayed call. Recall that you will be charged for incurred gas costs by the relay, so you should only accept calls that you're willing to pay for!
+
+There are multiple ways to make this work, including:
+ * having a whitelist of trusted users
+ * only accepting calls to an onboarding function
+ * charging users in tokens (possibly issued by you)
+ * delegating the acceptance logic off-chain
+
+All relayed call requests can be rejected at no cost to the recipient.
+
+=== pre and postRelayedCall
+
+You may have noticed one of the suggested acceptance flows involves charging a user in tokens. You may, for example, setup a system where users spend a special token to pay for relayed calls, and then either gift these to new users, or sell them off-chain. The GSN will let you easily implement these designs.
+
+After a relayed call is accepted, RelayHub will give your contract two opportunities to charge your user for their call, perform some bookeeping, etc.: _before_ and _after_ the actual relayed call is made. These functions are aptly named preRelayedCall and postRelayedCall.
+
+preRelayedCall will inform you of the maximum cost the call may have, and is used to charge the user in advance. This is only required in scenarios where your users may run off with their tokens before you're able to charge them, and not all applications need this function.
+
+postRelayedCall will give you an accurate estimate of the transaction cost, making it a natural place to charge users. It will also let you know if the relayed call reverted or not, which allows for more sophisticated interactions, like not charging users for reverted calls.
+
+== msg.sender and msg.data
+
+There's only one extra care you should take when working with GSN recipient contracts: _you must never use msg.sender or msg.data directly_. Understanding why requires an in-depth analysis of the inner workings of the GSN and RelayHub, but you need not worry about these low-level details: GSNRecipient provides two functions, `_msgSender()` and `_msgData()`, which are drop-in replacements for `msg.sender` and `msg.data`. As long as you use these two functions instead, you're good to go!
```

Comment:
> Understanding why requires an in-depth analysis of the inner workings of the GSN and RelayHub, but you need not worry about these low-level details

Let's replace this with a link to an explanation why this is the case. We don't need to go into details on how data and sender are actually serialized, just explain that the data & sender of a wrapped tx are not the same as those of the wrapper tx.